pkg/webhook/v1alpha2/pod/inject_webhook.go

```diff
@@ -0,0 +1,152 @@
+	JobRoleLabel = "job-role"
+)
+
+// For debug
+var log = logf.Log.WithName("injector-webhook")
+
+// sidecarInjector that inject metrics collect sidecar into master pod
+type sidecarInjector struct {
+	client         client.Client
+	decoder        types.Decoder
+	managerService string
+}
+
+var _ admission.Handler = &sidecarInjector{}
+
+func (s *sidecarInjector) Handle(ctx context.Context, req types.Request) types.Response {
+	pod := &v1.Pod{}
+	err := s.decoder.Decode(req, pod)
+	if err != nil {
+		return admission.ErrorResponse(http.StatusBadRequest, err)
+	}
+
+	// Check whether the pod need to be mutated
+	if !s.MutationRequired(pod) {
+		return admission.ValidationResponse(true, "")
+	}
+
+	// Get the namespace from req since the namespace in the pod is empty.
+	namespace := req.AdmissionRequest.Namespace
+	// Do mutation
+	mutatedPod, err := s.Mutate(pod, namespace)
+	if err != nil {
+		return admission.ErrorResponse(http.StatusBadRequest, err)
+	}
+
+	return admission.PatchResponse(pod, mutatedPod)
+}
+
+var _ inject.Client = &sidecarInjector{}
+
+func (s *sidecarInjector) InjectClient(c client.Client) error {
+	s.client = c
+	return nil
+}
+
+var _ inject.Decoder = &sidecarInjector{}
+
+func (s *sidecarInjector) InjectDecoder(d types.Decoder) error {
+	s.decoder = d
+	return nil
+}
+
+func NewSidecarInjector(c client.Client, ms string) *sidecarInjector {
+	return &sidecarInjector{
+		client:         c,
+		managerService: ms,
+	}
+}
+
+func (s *sidecarInjector) MutationRequired(pod *v1.Pod) bool {
+	value, err := s.GetLabel(pod, JobRoleLabel)
+	if err != nil || value != "master" {
+		return false
+	}
+	return true
+}
+
+func (s *sidecarInjector) GetLabel(pod *v1.Pod, targetLabel string) (string, error) {
+	labels := pod.Labels
+	for k, v := range labels {
+		if k == targetLabel {
+			return v, nil
+		}
+	}
+	return "", errors.New("Label " + targetLabel + " not found.")
+}
+
+func (s *sidecarInjector) Mutate(pod *v1.Pod, namespace string) (*v1.Pod, error) {
+	mutatedPod := pod.DeepCopy()
+
+	// Get the trial info from client
+	trialName, err := s.GetLabel(pod, JobNameLabel)
+	if err != nil {
+		return nil, err
+	}
+	trial := &trialsv1alpha2.Trial{}
+	err = s.client.Get(context.TODO(), apitypes.NamespacedName{Name: trialName, Namespace: namespace}, trial)
+	if err != nil {
+		return nil, err
+	}
+
+	metricName := trial.Spec.Objective.ObjectiveMetricName
+	for _, v := range trial.Spec.Objective.AdditionalMetricNames {
+		metricName += ";"
+		metricName += v
+	}
+
+	// Hard code container, inject metrics collector
+	injectContainer := v1.Container{
+		Name:            "metrics-collector",
+		Image:           "gcr.io/kubeflow-images-public/katib/v1alpha2/metrics-collector",
+		Command:         []string{"./metricscollector"},
+		Args:            []string{"-e", "TODO_Experiment", "-t", trialName, "-k", "TFJob", "-n", namespace, "-m", katibmanagerv1alpha2.GetManagerAddr(), "-mn", metricName},
```

Comment:
we can use "controller-name" label of Pod to get the "-k" value
we can get "-e" value, that is the experiment name, from trial's metadata ownerReferences field

Reply:
Done, got "-k" and "-e" values. It seems pytorch job does not have "controller-name" label yet?